File: service/app/src/main/java/io/miragon/blueprint/adapter/inbound/operaton/RequestCancellationWorker.java
package io.miragon.blueprint.adapter.inbound.operaton;

import dev.bpmcrafters.processengine.worker.ProcessEngineWorker;
import dev.bpmcrafters.processengine.worker.Variable;
import io.miragon.blueprint.adapter.process.CancelBikeOrderProcessApi.ServiceTasks;
import io.miragon.blueprint.adapter.process.CancelBikeOrderProcessApi.Variables;
import io.miragon.blueprint.application.port.inbound.RequestOrderCancellationUseCase;
import io.miragon.blueprint.domain.bike.OrderId;
import java.util.Map;
import org.springframework.stereotype.Component;

/**
 * Consumes the {@code requestCancellation} external task of the {@code cancelBikeOrder} sub-process. {@code orderId}
 * is handed in by the calling activity; the outcome ({@code cancellationPossible}) routes the gateway.
 */
@Component
public class RequestCancellationWorker {

    private final RequestOrderCancellationUseCase useCase;

    public RequestCancellationWorker(RequestOrderCancellationUseCase useCase) {
        this.useCase = useCase;
    }

    @ProcessEngineWorker(topic = ServiceTasks.REQUEST_CANCELLATION)
    public Map<String, Object> requestCancellation(@Variable String orderId) {
        boolean cancellationPossible = useCase.requestCancellation(new OrderId(orderId));
        return Map.of(Variables.ServiceTaskRequestCancellation.CANCELLATION_POSSIBLE.getValue(), cancellationPossible);
    }
}

File: service/app/src/main/kotlin/io/miragon/blueprint/adapter/inbound/operaton/RequestCancellationWorker.kt
package io.miragon.blueprint.adapter.inbound.operaton

import dev.bpmcrafters.processengine.worker.ProcessEngineWorker
import dev.bpmcrafters.processengine.worker.Variable
import io.miragon.blueprint.adapter.process.CancelBikeOrderProcessApi.ServiceTasks
import io.miragon.blueprint.adapter.process.CancelBikeOrderProcessApi.Variables
import io.miragon.blueprint.application.port.inbound.RequestOrderCancellationUseCase
import io.miragon.blueprint.domain.bike.OrderId
import org.springframework.stereotype.Component

/**
 * Consumes the `requestCancellation` external task of the `cancelBikeOrder` sub-process. `orderId`
 * is handed in by the calling activity; the outcome (`cancellationPossible`) routes the gateway.
 */
@Component
class RequestCancellationWorker(
    private val useCase: RequestOrderCancellationUseCase,
) {

    @ProcessEngineWorker(topic = ServiceTasks.REQUEST_CANCELLATION)
    fun requestCancellation(@Variable orderId: String): Map<String, Any?> {
        val cancellationPossible = useCase.requestCancellation(OrderId(orderId))
        return mapOf(Variables.ServiceTaskRequestCancellation.CANCELLATION_POSSIBLE.value to cancellationPossible)
    }
}
